### DISCUSSION TITLE
Can mongock help auto create index?

### DISCUSSION THREAD

**Discussion by h9ing** [flamingock/mongock#39] [Ideas]
## Description
Can Mongock help auto create index?

I clone sample code "mongock/mongock-examples/mongodb/springboot-quickstart" and run it.
I expect it can help me to create index, because the sample code have @CompoundIndex annotation, it can auto create index when i set "spring.data.mongodb.auto-index-creation=true" (now i set false when i use mongock).
But I don't see any index be created.

My question is should I write create index code in ChangeUnit? Or mongock can use @CompoundIndex/@Indexed(unique = true) to help me auto create index.

Thank you.

## PRIORITY
NORMAL

### Mongock
the same as sample code.
- Mongock version 5.1
- springboot(2.4.13)
- MongoDB(4.0.4)

## Steps to Reproduce

1. clone sample code from "https://github.com/mongock/mongock-examples/tree/master/mongodb/springboot-quickstart".
2. add @CompoundIndex and @Indexed(unique = true).
3. run

## Behaviour

**Expected behavior:** 
collection "clientCollection" have multiple index.

**Actual behavior:** 
collection "clientCollection" doesn't have any index.

**How often the bug happens:** 
every time

## Link to repository using Mongock
https://github.com/mongock/mongock-examples/tree/master/mongodb/springboot-quickstart

## Additional context
![example1](https://user-images.githubusercontent.com/51065108/170694259-ce6f52a6-ae6f-4f3c-990b-bdb894a85327.png)
<img width="474" alt="example2" src="https://user-images.githubusercontent.com/51065108/170694269-107fabf3-49c4-4f36-b287-fe8b2cf08752.png">

**Comment by osantana85** [comment]
Hello @9ing , currently Mongock doesn't manage it, but it sounds interesting for us and we will take it into account for a future release feature.

**Comment by h9ing** [comment]
HI @osantana85

Thanks for your reply.
I am looking forward to it release. 😃

**Comment by rfelgent** [comment]
I am facing the same issue and I created custom `ChangeUnit` java code. It feels somewhat duplicated code and violates the DRY principle.

It would be convinient, if mongock takes into account the annotations provided by spring-data.

My considerations: 

- Wouldn't this lead to a tight coupling of mongock to spring-data, would it ?
- Why using spring-data annotations for DDL stuff at all, if you use mongock for versioning (and eventually dropping spring-data's limited versioning capabilities?)

In the end: it might be smarter to introduce mongock-owned annotations (similar to spring-data ones) and add them the pojos rather than supporting spring-data annotations.

**Comment by dieppa** [comment]
Hello @rfelgent ,

very good questions.

- `Wouldn't this lead to a tight coupling of mongock to spring-data, would it ?`
No and yes. It wouldn't mean Mongock, as framework, being tightly coupled with spring-data. As you probably know, Mongock provides multiple drivers(MongoDB, DynamoDB, etc.) Within the MongoDB drivers set, there is one that is for spring-data(actually there are two: spring-data 2 and 3), so those drivers are the ones tightly coupled to spring-data(they are already!)

- `Why using spring-data annotations for DDL stuff at all, if you use mongock for versioning (and eventually dropping spring-data's limited versioning capabilities?)`
You got a point here. There are different point of views, all of them valid. I see your point, but from Mongock's point of view, the less intrusive Mongock is, the better. Adding Mongock's custom annotations for entity classes  would very intrusive in this sense. But I think it would be also nice to have some declarative mechanism for this kind of things, so you wouldn't need to create you own changeUnit for this.

In summary(for us and community to debate)
- Mongock team would like to provide a easy mechanism to do these things
- We want to keep simple and not intrusive
- We'd like to avoid reinventing the wheel

Also, @rfelgent , if you think that your code can help us and you are happy sharing, we would really appreciate it and of course you would get the credit and mentioning!
